notebook/notebook/notebook.py

```diff
@@ -113,201 +143,334 @@ def start_pod(jupyter_token, image):
                 'app': 'notebook-worker',
                 'hail.is/notebook-instance': INSTANCE_ID,
                 'uuid': pod_id,
-            }),
+                'svc_name': svc.metadata.name,
+                **labels
+            },),
         spec=pod_spec)
     pod = k8s.create_namespaced_pod(
         'default',
         pod_template,
-        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
+        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS,
+    )
+
     return svc, pod
 
 
-def external_url_for(path):
-    # NOTE: nginx strips https and sets X-Forwarded-Proto: https, but
-    # it is not used by request.url or url_for, so rewrite the url and
-    # set _scheme='https' explicitly.
-    protocol = request.headers.get('X-Forwarded-Proto', None)
-    url = flask.url_for('root', _scheme=protocol, _external='true')
-    return url + path
+def del_pod(pod_name: str):
+    k8s.delete_namespaced_pod(
+        pod_name,
+        'default',
+        kube.client.V1DeleteOptions(),
+        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
+
+
+def del_svc(svc_name: str):
+    k8s.delete_namespaced_service(
+        svc_name,
+        'default',
+        kube.client.V1DeleteOptions(),
+        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
+
+
+# General resource marshalling functions #
+
+
+def get_path(data, path: str):
+    # Not using tail recursion because python doesn't optimize such calls
+    while True:
+        idx = path.find('.')
+
+        if idx == -1:
+            if isinstance(data, dict):
+                return data[path]
+            return getattr(data, path)
+
+        data = getattr(data, path[0:idx])
 
+        path = path[idx + 1:]
+
+
+def marshall_json(resources: [], paths=[], flatten=False):
```

Comment:
Can you help me understand the value to testing, understandability, correctness, or security of this method?

A method like this prevents duplicate code but has a self-descriptive name and is very easy to understand.

```
def k8s_pod_to_json(pod, svc_json=None):
    if svc_json is None:
        svc_json = svc_to_json(read_svc(... pod.metadata.labels['svc_name'] ...))
    return {'name': pod.metadata.labels['name'],
            'pod_name': pod.metadata.name,
            ...}
# ...
return k8s_pod_to_json(pod)
```

I'm a bit dubious on making network requests in something called `k8s_pod_to_json` (or `marshall_json`), but 🤷‍♀️, I'm not that stuck on it.

Reply:
Could you address your specific concerns with this function? If the goal is to make future improvements, I think you should make that an issue, assign it. If there is something here that prevents correct function, lets address that.